src/config.js

```diff
@@ -521,6 +525,59 @@ export default class Config {
     }
   }
 
+  async findManifest(dir: string, isRoot: boolean): Promise<?Manifest> {
+    for (const registry of registryNames) {
+      const manifest = await this.tryManifest(dir, registry, isRoot);
+
+      if (manifest) {
+        return manifest;
+      }
+    }
+
+    return null;
+  }
+
+  async findWorktree(initial: string): Promise<?string> {
+    let previous = null;
+    let current = path.normalize(initial);
+
+    do {
+      const manifest = await this.findManifest(current, true);
+
+      if (manifest && manifest.workspaces) {
+        return current;
+      }
+
+      previous = current;
+      current = path.dirname(current);
+    } while (current !== previous);
+
+    return null;
+  }
+
+  async resolveWorkspaces(root: string, patterns: Array<string>): Promise<{ [string]: { loc: string, manifest: Manifest } }> {
+    const compiledPatterns = filter.ignoreLinesToRegex(patterns);
+
+    // We need the ignoreFiles, not the keepFiles, because the patterns are exclusion patterns rather than the opposite
+    const ignoreBasenames = new Set(this.registryFolders);
+    const files = filter.sortFilter(await fs.walk(root, undefined, ignoreBasenames), compiledPatterns).ignoreFiles;
```

Comment:
The code above is also using the glob syntax, because of `sortFilters`. I tried to use the same implementation from what we use for `pack`, but maybe `glob` is more efficient, yeah.